Feature/syntax highlighting

richardbuckle opened the pull request:
Fixes #1032

richardbuckle commented:
I removed commments like `//end if` on blocks becase our modern tooling makes then redundant, and they are prone to get out of date.

Tkael commented:
Initial impression is very positive. :-) 
1. Spacing in the header for the edit window is perhaps more generous than needed?
2. The script description textbox appears to be MIA?
3. Script comparison doesn't appear to be functioning correctly? No highlights show the ")" sign I deleted in line 10 below.
![Untitled](https://user-images.githubusercontent.com/4382539/70127070-28159b00-162f-11ea-9793-54de01e22f3e.png)
4. Tooltips for the `Delete` and the `Delete line` buttons both describe the respective buttons as simply `Delete`. These should perhaps be revised to `Delete selection` and `Delete line`.
5. The `Test script` button seems to be referencing something other than the current script. If I delete the full contents of the `AFMU repairs` script and replace it with just `{if event.item:` (which should always throw an exception), the vocal response is `Cargo hatch fully repaired`.

Tkael commented:
6. Should we continue to use a grey background for the diff view (white suggests that it is editable when it is not)?

Tkael commented:
Note: For item 5 above, please ref. EditScriptWindow.xaml.cs line 155.
The test script is rendered incorrectly using `ScriptValue` but it would be rendered correctly if `scriptView.Text` were used instead.
I suspect item 3 may need a similar tweak?

richardbuckle commented:
1. Take another look now that the script description is back. Thoughts welcomed.
2. Good catch, fixed.
3. Fixed with 5 by reading the current value of `scriptView.Text`.
4. Fixed. Kept the former *as is* for consistency with cut, copy and paste, but I'm open to persuasion. These are of course localisable as you probably saw.
5. Fixed with 2.
6. We were never consistent here: the script view window has always used a white background. TBH I've never been a fan of the grey background and didn't clock that it was trying to indicate read-only; I guess I kind of assumed it was a dev oversight. Thoughts welcomed.

richardbuckle commented:
Addendum: trying out background colors of "White" for the editable view and "WhiteSmoke" for the read-only views.

richardbuckle commented:
Thinking aloud: I routinely see read-only documents with a white background, for example web pages and PDFs. I'm not sold on a grey background indicating read-only status -- for my part I merely assumed that it was either an oversight or some random technical limitaion.

Tkael commented on SpeechResponder/EditScriptWindow.xaml:
```diff
@@ -22,11 +97,20 @@
             <TextBox Margin="10" Grid.Row="0" Grid.Column="1" Text="{Binding Path=ScriptName, Mode=TwoWay}" IsReadOnly="{Binding Path=Responder}"/>
-            <TextBlock Margin="10" Grid.Row="1" Grid.Column="0" Text="{x:Static resx:SpeechResponder.header_description}"></TextBlock>
+            <TextBlock Margin="10" Grid.Row="1" Grid.Column="0" Text="{x:Static resx:SpeechResponder.header_description}"/>
             <TextBox Margin="10" Grid.Row="1" Grid.Column="1" Text="{Binding Path=ScriptDescription, Mode=TwoWay}" IsReadOnly="{Binding Path=Responder}"/>
```
Consider using `Margin="10, 5, 10, 0"` for lines 100-103 (more compact).

richardbuckle replied:
Done

Tkael commented on SpeechResponder/ViewScriptWindow.xaml:
```diff
@@ -25,7 +26,16 @@
             <TextBlock Margin="10" Grid.Row="0" Grid.Column="1" Text="{Binding Path=ScriptName}"/>
             <TextBlock Margin="10" Grid.Row="1" Grid.Column="0" Text="{x:Static resx:SpeechResponder.header_description}" />
             <TextBlock Margin="10" Grid.Row="1" Grid.Column="1" Text="{Binding Path=ScriptDescription}"/>
```
Consider using `Margin="10, 5, 10, 0"` for lines 25-28 (more compact).

richardbuckle replied:
Done

Tkael commented on SpeechResponder/EditScriptWindow.xaml:
```diff
@@ -10,6 +11,80 @@
         ResizeMode="CanResizeWithGrip"
         Height="600" Width="800">
     <DockPanel LastChildFill="True" Background="#FFE5E5E5">
+        <ToolBar DockPanel.Dock="Top" Background="#FFE5E5E5">
```
The "grip" on the left side of the toolbar can be hidden by adding `ToolBarTray.IsLocked="True"`.

richardbuckle replied:
Done

Tkael commented on SpeechResponder/ShowDiffWindow.xaml:
```diff
@@ -3,17 +3,21 @@
+                           FontFamily="Consolas"
+                           IsReadOnly="True"
+                           WordWrap="True"
+                           ShowLineNumbers="True"
```
At present, the line numbers in the diff window do not match the line numbers in the edit window.
Would it be possible to adjust line numbers so that the diff window preserves the same line numbers as the edit window (perhaps by repeating the line number or omitting a repeated line number while displaying a diff)?

The diff window also adds a trailing line that isn't present in the original text.

![Untitled](https://user-images.githubusercontent.com/4382539/70199976-fdc3ec00-16c7-11ea-8e04-be4f7d611933.png)
Note: This screenshot incorporates some recommendations on margins and the toolbar grip discussed elsewhere in my comments.

richardbuckle replied:
Unfortunately line numbers in AvalonEdit are an integral part of the text engine and can't be set arbitrarily. I think it's simplest to just turn off line number display in the diff window. I was adding an extra newline after the last line segment: fixed.

Tkael commented on SpeechResponder/SpeechResponder.cs:
```diff
@@ -34,6 +34,8 @@ public class SpeechResponder : EDDIResponder
         protected static List<Event> eventQueue = new List<Event>();
         private static readonly object queueLock = new object();
 
+        private readonly CottleHighlightingDefinition cottleHighlightingDefinition = new CottleHighlightingDefinition();
```
This appears to be unused. Dead code?

richardbuckle replied:
Instantiating this object installs the Cottle grammar definition into AvalonEdit. I'll also be using it to load and save any customisation by the user.

Tkael replied:
Ok. My code analysis identified it as dead code so it might be worth dropping in a comment stating why it's needed.

Tkael commented on Tests/GeneratorTests.cs:
```diff
@@ -101,5 +100,23 @@ public void TestGenerateWikiEventsList()
+            string output = "      <Word>" + string.Join("</Word>\r\n      <Word>", eventVars) + " </Word>\r\n";
+            Directory.CreateDirectory(@"Cottle\");
+            File.WriteAllText(@"Cottle\Custom keywords.txt", output);
+        }
```
Should this file include a header with instructions? Something like "Paste the generated text below into the "Custom properties" rule set in `SpeechResponder\Cottle.xshd`"?

Tkael replied:
It's probably not very possible without revamping how we define custom functions in Cottle, but I assume you also considered generating a list of custom functions?

richardbuckle replied:
I think I'll put instructions in a comment rather than in the output.

I did look into generating the list of custom functions but it would have involved more restructuring than I was willing to take on for this iteration. So I just extracted the names manually using Notepad++ and some creative use of regex. I still hope to do it though.

Tkael replied:
One major change at a time? Fair enough. 😉
